File: src/lib/normalize-brackets.ts
// Normalize brackets in filenames with context awareness
// This function attempts to intelligently fix bracket imbalances

// Known tag patterns that should be self-contained in brackets
export const KNOWN_TAGS = [
  'DL版',
  '中文翻訳',
  '中国翻訳',
  '無修正',
  '進行中',
  '完結',
  'FANZA特別版',
  'デジタル版',
  '単行本版',
  '修正版',
]

// Create regex pattern for known tags
const TAG_PATTERN = new RegExp(`\\[(${KNOWN_TAGS.join('|')})(?!\\])`, 'g')

export default function normalizeBrackets(filename: string): string {
  let normalized = filename

  // First, try to fix known tag patterns
  // Example: [DL版 (オリジナル) -> [DL版] (オリジナル)
  normalized = normalized.replace(TAG_PATTERN, (match, tag) => {
    // Check what comes after the tag
    const afterTag = normalized.substring(normalized.indexOf(match) + match.length)

    // If followed by a space and opening bracket, close the tag bracket first
    if (/^\s*[\(\[\{]/.test(afterTag)) {
      return `[${tag}]`
    }

    return match
  })

  // Count remaining bracket imbalances
  const brackets = [
    { open: '(', close: ')' },
    { open: '[', close: ']' },
    { open: '{', close: '}' },
    { open: '（', close: '）' },
    { open: '［', close: '］' },
    { open: '｛', close: '｝' },
  ]

  for (const bracket of brackets) {
    const { open, close } = bracket
    const openCount = (normalized.match(new RegExp(`\\${open}`, 'g')) || []).length
    const closeCount = (normalized.match(new RegExp(`\\${close}`, 'g')) || []).length

    if (openCount > closeCount) {
      // More opening brackets - add closing brackets at the end
      normalized += close.repeat(openCount - closeCount)
    }
  }

  return normalized
}

File: src/helpers/normalize_brackets.rs
pub const KNOWN_TAGS: &[&str] = &[
    "DL版",
    "中文翻訳",
    "中国翻訳",
    "無修正",
    "進行中",
    "完結",
    "FANZA特別版",
    "デジタル版",
    "単行本版",
    "修正版",
];

pub fn normalize_brackets(filename: &str) -> String {
    let mut normalized = String::from(filename);

    // Fix known tag patterns without closing bracket:
    // [DL版 (オリジナル) -> [DL版] (オリジナル)
    for tag in KNOWN_TAGS {
        let open_pattern = format!("[{tag}");
        let closed_pattern = format!("[{tag}]");
        // Only fix if the tag appears without a closing bracket
        // i.e., "[DL版" exists but "[DL版]" does not at that position
        loop {
            let Some(pos) = normalized.find(&open_pattern) else {
                break;
            };
            let after_tag = pos + open_pattern.len();
            // Check if the character right after the tag is already ']'
            if normalized[after_tag..].starts_with(']') {
                // Already properly closed — skip this occurrence
                // To avoid infinite loop, we need to move past this match
                break;
            }
            // Check if followed by whitespace + bracket
            let after = normalized[after_tag..].trim_start();
            if after.starts_with('(') || after.starts_with('[') || after.starts_with('{') {
                normalized = format!(
                    "{}{}{}",
                    &normalized[..pos],
                    closed_pattern,
                    &normalized[after_tag..]
                );
            } else {
                break;
            }
        }
    }

    // Count remaining bracket imbalances
    let brackets: &[(&str, &str)] = &[
        ("(", ")"),
        ("[", "]"),
        ("{", "}"),
        ("（", "）"),
        ("［", "］"),
        ("｛", "｝"),
    ];

    for &(open, close) in brackets {
        let open_count = normalized.matches(open).count();
        let close_count = normalized.matches(close).count();
        if open_count > close_count {
            for _ in 0..(open_count - close_count) {
                normalized.push_str(close);
            }
        }
    }

    normalized
}
